core/src/main/java/com/microsoft/applicationinsights/channel/concrete/inprocess/PartialSuccessHandler.java

```diff
@@ -0,0 +1,196 @@
+package com.microsoft.applicationinsights.channel.concrete.inprocess;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.zip.GZIPInputStream;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.http.HttpStatus;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.common.base.Optional;
+import com.microsoft.applicationinsights.internal.channel.TransmissionHandler;
+import com.microsoft.applicationinsights.internal.channel.TransmissionHandlerArgs;
+import com.microsoft.applicationinsights.internal.channel.common.GzipTelemetrySerializer;
+import com.microsoft.applicationinsights.internal.channel.common.Transmission;
+import com.microsoft.applicationinsights.internal.channel.common.TransmissionPolicyManager;
+import com.microsoft.applicationinsights.internal.logger.InternalLogger;
+
+/**
+ * This class implements the retry logic for partially accepted transmissions.
+ * HTTP status code 206.
+ * <p>
+ * 
+ * @see <a href=
+ *      "https://github.com/Microsoft/ApplicationInsights-dotnet/blob/master/docs/ServerTelemetryChannel%20error%20handling.md#partialsuccesstransmissionpolicy">PartialSuccessTransmissionPolicy</a>
+ * @author jamdavi
+ *
+ */
+public class PartialSuccessHandler implements TransmissionHandler {
+
+	/**
+	 * Ctor
+	 * 
+	 * Constructs the PartialSuccessHandler object.
+	 * 
+	 * @param policy
+	 *            The {@link TransmissionPolicyManager} object that is needed to
+	 *            control the back off policy.
+	 */
+	public PartialSuccessHandler(TransmissionPolicyManager policy) {
+	}
+
+	@Override
+	public void onTransmissionSent(TransmissionHandlerArgs args) {
+		validateTransmissionAndSend(args);
+	}
+
+	/**
+	 * Provides the core logic for the retransmission
+	 * 
+	 * @param args
+	 *            The {@link TransmissionHandlerArgs} for this transmission.
+	 * @return Returns a pass/fail for handling this transmission.
+	 */
+	boolean validateTransmissionAndSend(TransmissionHandlerArgs args) {
+		if (args.getTransmission() != null && args.getTransmissionDispatcher() != null) {
+			switch (args.getResponseCode()) {
+			case HttpStatus.SC_PARTIAL_CONTENT:
+				BackendResponse beR = getBackendResponse(args.getResponseBody());
+				List<String> originalItems = generateOriginalItems(args);
+
+				// Somehow the amount of items received and the items sent do not match
+				if (beR != null && (originalItems.size() != beR.itemsReceived)) {
+					InternalLogger.INSTANCE.trace(
+							"Skipping partial content handler due to itemsReceived being larger than the items sent.");
+					return false;
+				}
+
+				if (beR != null && (beR.itemsAccepted != beR.itemsReceived)) {
+
+					List<String> newTransmission = new ArrayList<String>();
+					for (BackendResponse.Error e : beR.errors) {
```

Comment:
Why don't we replace this for-each loop with a for loop like this : 
for (int i = 0; i < ber.errors.length; ++i). This would help simplify the index operation being used mutiple times. We can the just do 
if (i < originalItems.size()) {
     newTransmission.add(originalItems.get(i));
}

Reply:
The index of the items in the errors array is not guaranteed to be the index of the original item sent.

errors[0] = Error() { index = 2, responseCode = 500, message = "Internal Error" }

Reply:
Ah this is a catch. Got it